libraries/chain/include/eosio/chain/fork_database.hpp

```diff
@@ -95,15 +98,86 @@ namespace eosio::chain {
+    */
+   class fork_database {
+      mutable std::recursive_mutex m;
+      const std::filesystem::path data_dir;
+      std::variant<fork_database_t<block_state_legacy_ptr>, fork_database_t<block_state_ptr>> v;
+   public:
+      explicit fork_database(const std::filesystem::path& data_dir);
+      ~fork_database(); // close on destruction
+
+      void open( validator_t& validator );
+
+      void switch_from_legacy();
+
+      // see fork_database_t::fetch_branch(forkdb->head()->id())
+      std::vector<signed_block_ptr> fetch_branch_from_head();
+
+      template <class R, class F>
+      R apply(const F& f) {
+         std::lock_guard g(m);
+         if constexpr (std::is_same_v<void, R>)
+            std::visit([&](auto& forkdb) { f(forkdb); }, v);
+         else
+            return std::visit([&](auto& forkdb) -> R { return f(forkdb); }, v);
+      }
+
+      template <class R, class F>
+      R apply(const F& f) const {
+         std::lock_guard g(m);
+         if constexpr (std::is_same_v<void, R>)
+            std::visit([&](const auto& forkdb) { f(forkdb); }, v);
+         else
+            return std::visit([&](const auto& forkdb) -> R { return f(forkdb); }, v);
+      }
+
+      template <class R, class F>
+      R apply_if(const F& f) {
+         if constexpr (std::is_same_v<void, R>)
+            std::visit(overloaded{[&](fork_database_legacy_t&) {},
+                                  [&](fork_database_if_t& forkdb) {
+                                     std::lock_guard g(m);
+                                     f(forkdb);
+                                  }}, v);
+         else
+            return std::visit(overloaded{[&](fork_database_legacy_t&) -> R { return {}; },
+                                         [&](fork_database_if_t& forkdb) -> R {
+                                            std::lock_guard g(m);
+                                            return f(forkdb);
+                                         }}, v);
+      }
+
+      template <class R, class F>
+      R apply_dpos(const F& f) {
+         if constexpr (std::is_same_v<void, R>)
+            std::visit(overloaded{[&](fork_database_legacy_t& forkdb) {
+                                     std::lock_guard g(m);
+                                     f(forkdb);
+                                  },
+                                  [&](fork_database_if_t&) {}}, v);
+         else
+            return std::visit(overloaded{[&](fork_database_legacy_t& forkdb) -> R {
+                                            std::lock_guard g(m);
+                                            return f(forkdb);
+                                         },
+                                         [&](fork_database_if_t&) -> R {
+                                            return {};
+                                         }}, v);
+      }
+
+      // if we every support more than one version then need to save min/max in fork_database_t
```

Comment:
Comment doesn't read right. Maybe this
```suggestion
      // if we ever support more than one version, then we need to save min/max in fork_database_t
```